File: counting_valleys/CountingValleys.js
function countingValleys(steps, path) {
    // Write your code here
    let prev_level = 0;
    let current_level = 0;
    let valleys = 0;
    let path_arr = path.split('');

    path_arr.forEach(step => {
        if (step === 'D') {
            prev_level = current_level;
            current_level -= 1;
        } else if (step === 'U') {
            prev_level = current_level;
            current_level += 1;
            if (prev_level < 0 && current_level >= 0) valleys += 1;
        }
    });
    return valleys;
}

File: counting_valleys/counting_valleys.rb
def countingValleys(_steps, path)
  # Write your code here
  current_sea_level = 0
  prev_sea_level = 0
  valleys_traversed = 0

  path.each_char do |step|
    if step == 'D'
      prev_sea_level = current_sea_level
      current_sea_level -= 1
    elsif step == 'U'
      prev_sea_level = current_sea_level
      current_sea_level += 1
      valleys_traversed += 1 if current_sea_level >= 0 && prev_sea_level < 0
    end
  end
  valleys_traversed
end